### DISCUSSION TITLE
How to get all the nodes of an editor's state?

### DISCUSSION THREAD

**Discussion by JohnnyRacer** [facebook/lexical#4774] [Q&A]
Hello, is there a way of getting all the nodes as a tree of `LexicalNode` objects similar to how `editor.toJSON()` dumps the entire structure of the editor to a JSON object? I can only seem to find documentation on getting a list of nodes of a selection via `getNodes()`.

**Comment by mariusplanable** [comment]
I needed something similar but only for the current selection, and used a hack with `@lexical/html`.
I generate the HTML from selection, then get back the nodes from the HTML and get the array of top level nodes from the selection:
```typescript
const html = $generateHtmlFromNodes(editor, selection);
const dom = new DOMParser().parseFromString(html, 'text/html');
const nodesFromDOM = $generateNodesFromDOM(editor, dom);
```
then I iterate through these nodes and recursively through the children of each node and create the serialized structure I need.
```typescript
function buildRecursivelySerializedNode(node) {
  if ($isElementNode(node)) {
    return {
      ...node.exportJSON(),
      children: node.getChildren().map(buildRecursivelySerializedNode)
    };
  }

  return node.exportJSON();
}

// ...

const serializedState = nodesFromDOM.map(buildRecursivelySerializedNode);
```

**Reply by JohnnyRacer** [reply]
That's a cool hack, but I think it might be a bit slow for larger documents as it needs to serialize/deserialize to HTML for the entire tree though right? I think it might be a lot faster just to dump the entire document to JSON first using `editor.getEditorState().toJSON()` and then iterate through the node types and recreate the nodes using their constructors. Also I believe you can just do the following to get the nodes from a selection instead of dealing with any serialization at all like so:
```js

editor.read(() => {
  const selection = $getSelection();
  const nodes = selection.getNodes();
 // Do something with the nodes 
}); // Invoke the function when you need to get nodes of selection

```

**Comment by Michael-indochat** [comment]
Excuse me, I am implementing to filter out the text length of mentionNode,
What I am currently thinking of is to use `getChildren` to get all nodes and judge each node. If any node is menitonNode, I need to filter it out.

At present, I have not thought of the correct implementation direction. Can you give me some suggestions or template references?

**Comment by Alex-ray** [comment]
You can use [`$nodesOfType`](https://lexical.dev/docs/api/modules/lexical#nodesoftype)!

```
import { $nodesOfType } from "lexical";

editorState.read(() => {
    const listNodes = $nodesOfType(ListNode);
})
```

**Reply by CanRau** [reply]
Though this will only give you nodes of a specific type, not just all nodes.

Hoped I could do `$nodesOfType(LexicalNode)`, unfortunately `LexicalNode` is only exported as `type` and `$nodesOfType` wouldn't accept it as its argument needs to `extend LexicalNode` and probably wouldn't work anyway 😅 

I'd love something like this, as my custom `OnChange` plugin also gives me some "statistics" on the editor state.

Just adapted `getAllTextNodes` and it works pretty good so far

```ts
import type { LexicalNode, ElementNode } from "lexical";
import { $isElementNode } from "lexical";

export function $getAllNodes(root: ElementNode): Array<LexicalNode> {
  const nodes = [];
  let child: LexicalNode | null = root.getFirstChild();
  while (child !== null) {
    nodes.push(child);
    if ($isElementNode(child)) {
      const subChildrenNodes = $getAllNodes(child);
      nodes.push(...subChildrenNodes);
    }
    child = child.getNextSibling();
  }
  return nodes;
}

// if you only need the number of nodes overall
export function $getAllNodesCount(root: ElementNode) {
  let nodeCount = 0;
  let child: LexicalNode | null = root.getFirstChild();
  while (child !== null) {
    nodeCount++;
    if ($isElementNode(child)) {
      nodeCount += $getAllNodesCount(child);
    }
    child = child.getNextSibling();
  }
  return nodeCount;
}
```

used like so

```ts
editorState.read(() => { // or editorState.update
  const root = $getRoot();
  const nodeCount = $getAllNodesCount(root); // could be any other parent node
});
```

Lemme now if there's a better or existing way I missed 🙏🏼 

Update: just now found `$getChildrenRecursively` though as I'm currently actually only interested in the count I guess my `$getAllNodesCount` should be "faster" as it doesn't load all nodes into memory, though `$getChildrenRecursively` would probably replace my `$getAllNodes` although the implementations are slightly different.

doh 🤦🏼 `$dfs` might also be interesting to get all nodes (again loading them into memory, if that's what you need), which also gives you the `depth` along with the `node`.

**Reply by girish-kadambari** [reply]
Hi @CanRau  I'm not geting root.getFirstChild(); funcrion it is telling there is no such property for LexicalNode

Property 'getFirstChild' does not exist on type 'LexicalNode'.ts(2339)
any

**Reply by CanRau** [reply]
Hmm interesting, not sure if this was already like that when I posted the code or some recent change, I'll have to look into it when I find the time.

Is this just TS complaining or does it actually not work? The latter would suggest some API changes I've missed 🤔

**Reply by CanRau** [reply]
Update the code to use `ElementNode` in my case it works and shouldn't make a difference other than fixing the TS error (@girish-kadambari)

**Comment by Julhol-droid** [comment]
I think you could also use the Lexical Utils Package with the $dfs function
https://lexical.dev/docs/api/modules/lexical_utils#dfs

**Reply by CanRau** [reply]
yep depending on use-case would also prefer $dfs, I briefly mentioned it at the end of a former [comment](https://github.com/facebook/lexical/discussions/4774#discussioncomment-7998448)
